### ISSUE TITLE
SQL Server fails to create database

### ISSUE THREAD

**Issue by lowleveldesign** [billziss-gh/winfsp#240]
## Bug Report

Creating a SQL Server LocalDB database fails on the memory file system (memfs-dotnet) when the destination file is in a subdirectory. Works OK when you save the file to the root folder on the volume. 

The error is:

> Msg 5177, Level 16, State 1, Server DESKTOP-TEST\LOCALDB#513B787F, Line 1                                                                                    An unexpected error occurred while checking the sector size for file 'x:\test\testdb.mdf'. Move the file to a local NTFS volume, where the sector size can be retrieved. Check the SQL Server error log for more information.
Msg 1802, Level 16, State 4, Server DESKTOP-TEST\LOCALDB#513B787F, Line 1
CREATE DATABASE failed. Some file names listed could not be created. Check related errors.

I'm attaching a file for the success (saving to the root folder) and the failure (saving to the test subfolder).

### How to Reproduce

1. Download and install SQL Server LocalDB (https://www.microsoft.com/en-us/sql-server/sql-server-downloads - it's a part of the Express installation). You may also need sqlcmd to connect to the db (from https://docs.microsoft.com/en-us/sql/tools/sqlcmd-utility?view=sql-server-2017) but any SQL Server client is fine.

2. Start memfs-dotnet: `memfs_dotnet.exe -d -1 -D c:\logs\memfs.log -m X:` (make sure C:\logs folder exists)

3. Create an empty folder on X: volume: `mkdir X:\test`

4. Create a SQL Server LocalDB instance and try to create a database:

```
> SqlLocalDB.exe c "test-fs" 
> SQLCMD.EXE -S "(localdb)\test-fs" -Q "create database test on primary (name=testdb, filename='x:\test\testdb.mdf') log on (name=testdb_lob, filename='C:\logs\testdb.ldf')"
```

The error should occur. Now, try the same but replace `x:\test\testdb.mdf` with `x:\testdb.mdf` and the call will succeed.

### Behaviors

The expected behavior is the creation of the .mdf file.

### Environment

- OS version and build: 10.0.18362 Build 18362 (Windows 10 Pro)
- WinFsp version and build: tested on 2019.2 and 2019.3 B2

[memfs-failure.log](https://github.com/billziss-gh/winfsp/files/3457134/memfs-failure.log)
[memfs-success.log](https://github.com/billziss-gh/winfsp/files/3457135/memfs-success.log)

**Comment by billziss-gh** [comment]
Sebastian, thanks. I will have a look. Please ping me directly if this is critical.

> Msg 5177, Level 16, State 1, Server DESKTOP-TEST\LOCALDB#513B787F, Line 1 An unexpected error occurred while checking the sector size for file 'x:\test\testdb.mdf'. Move the file to a local NTFS volume, where the sector size can be retrieved. Check the SQL Server error log for more information.
Msg 1802, Level 16, State 4, Server DESKTOP-TEST\LOCALDB#513B787F, Line 1
CREATE DATABASE failed. Some file names listed could not be created. Check related errors.

This error is quite interesting. The sector size is what is set in `VolumeParams` via the [`SectorSize`](https://github.com/billziss-gh/winfsp/blob/v1.5B2/src/dotnet/FileSystemHost.cs#L84) property when the file system is first created. (And of course the "sector size" is misnamed for non-disk file systems, but it determines the granularity at which non-buffered I/O is done.)

The sector size is always the same for all files whether they are in the root directory or a subdirectory. So I find it perplexing that SQL Server fails on files in a subdirectory but not in the root directory. It may be that SQL Server issues an unknown IOCTL to the file system in some cases, which WinFsp does not currently handle.

BTW, I am assuming that this problem happens with native MEMFS as well? Also do you happen to have FileSpy logs?

**Comment by kenjioba** [comment]
Bill, Sebastian is out on vacation but he did say that the issue immediately reproduces on the MEMFS driver as well.

**Comment by lowleveldesign** [comment]
Hi Bill, this problem happens on the native MEMFS as well. Please find attached the FileSpy logs for the successful and failed request.
[filespy-logs.zip](https://github.com/billziss-gh/winfsp/files/3514722/filespy-logs.zip)

**Comment by billziss-gh** [comment]
Sorry for the long delay in responding. I have been on a long vacation with sporadic access to the Internet. I am catching up on emails and GitHub issues, but I hope to have some feedback on this soon.

**Comment by billziss-gh** [comment]
I looked at the logs and it appears that the culprit **MAY** be the `IOCTL_MOUNTDEV_QUERY_DEVICE_NAME` IOCTL that SQL server insists on sending to the file system. WinFsp fails the IOCTL in both cases ("success" and "failure"), but perhaps this only affects SQL Server in the case where the IOCTL is sent to a directory (`\test`).

According to the [docs](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/mountmgr/ni-mountmgr-ioctl_mountdev_query_device_name) this IOCTL is sent by the Mount Manager, which WinFsp does not use. So it is curious that SQL Server sends it at all.

Does the problem persist if the file system is mounted as a network drive?

**Comment by lowleveldesign** [comment]
Hmm that's interesting. I tried: `.\memfs-x64.exe -u \ts\share` and then `SQLCMD.EXE -S "(localdb)\test-fs" -Q "create database test on primary (name=testdb, filename='\\ts\share\test\testdb.mdf') log on (name=testdb_lob, filename='C:\logs\testdb.ldf')"` and it worked. So it seems to be working when mounted as a network drive. 

But switching to a network drive would not allow us to use directory as a mount point (for example, T:\OneDrive), right?

**Comment by billziss-gh** [comment]
There is also the very curious `STATUS_OBJECT_NAME_INVALID` that appears on both "success" and "failure" logs. The file names where this error code appears look perfectly reasonable, so I am not sure what is going on there.

**Comment by billziss-gh** [comment]
> Hmm that's interesting. I tried: .\memfs-x64.exe -u \ts\share and then SQLCMD.EXE -S "(localdb)\test-fs" -Q "create database test on primary (name=testdb, filename='\\ts\share\test\testdb.mdf') log on (name=testdb_lob, filename='C:\logs\testdb.ldf')" and it worked. So it seems to be working when mounted as a network drive.

Interesting indeed. This gives more credence to the theory that this is somehow related to the MountManager IOCTL's.

> But switching to a network drive would not allow us to use directory as a mount point (for example, T:\OneDrive), right?

I believe you are correct.

**Comment by billziss-gh** [comment]
BTW, one other thing that I noticed from the logs is that SQL Server opens all files with the case-sensitive flag which is unusual (although perhaps usual for SQL Server):

- Are you running the file system as a case-insensitive file system (i.e. the `-i` flag in MEMFS)?
- Is this by any chance on a system that has the system-wide case-sensitive flag on (e.g. see this [link](https://superuser.com/questions/266110/how-do-you-make-windows-7-fully-case-sensitive-with-respect-to-the-filesystem)).

**Comment by lowleveldesign** [comment]
- Adding the -i flag did not change anything (same error). 
- The system has the default case-insensitive setting (obcaseinsensitive = 1).

I noticed that in the failed scenario SQL Server never makes the  `IRP_MJ_QUERY_VOLUME_INFORMATION` request. I also recorded [a filespy trace for an NTFS portable drive](https://github.com/billziss-gh/winfsp/files/3519959/portable-ntfs-success.spy.txt). The initial requests do not look that much different. There are a few `IRP_MJ_FILE_SYSTEM_CONTROL/IRP_MN_USER_FS_REQUEST	00060870` requests to which WinFsp returns `STATUS_INVALID_DEVICE_REQUEST` while NTFS sends `STATUS_NOT_A_REPARSE_POINT`, but that would be a strange reason for failure.

**Comment by billziss-gh** [comment]
> I also recorded a filespy trace for an NTFS portable drive.

Thanks for that last log. Very useful! This shows that even on an NTFS drive, the same failures happen:

```
IRP_MJ_FILE_SYSTEM_CONTROL      E:\test	                STATUS_NOT_A_REPARSE_POINT	FSCTL_GET_REPARSE_POINT (000900A8)

FASTIO_DEVICE_CONTROL           E:\test	                FAILURE	                        IOCTL_MOUNTDEV_QUERY_DEVICE_NAME (004D0008)
IRP_MJ_DEVICE_CONTROL           E:\test	                STATUS_INVALID_PARAMETER	IOCTL_MOUNTDEV_QUERY_DEVICE_NAME (004D0008)

IRP_MJ_CREATE                   E:\test\testdb.mdf	STATUS_OBJECT_NAME_INVALID	FILE_OPEN CreOpts: 00200020 Access: 00100080 Share:  00000003 Attrib: 00000080
```

- The `FSCTL_GET_REPARSE_POINT` failure makes sense as there is no reparse point here.
- The `IOCTL_MOUNTDEV_QUERY_DEVICE_NAME` failures are the same as in the MEMFS logs.
- The `STATUS_OBJECT_NAME_INVALID` failure is the same as in MEMFS again.

So in fact MEMFS and NTFS appear to behave the same way in these logs.

> I noticed that in the failed scenario SQL Server never makes the `IRP_MJ_QUERY_VOLUME_INFORMATION` request.

Good catch. This is where it would get the block size information (e.g. via the Win32 [`GetDiskFreeSpaceW`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getdiskfreespacew) API).

----

Silly idea. Have you tried passing the `-F NTFS` flag to MEMFS. Maybe SQL wants NTFS by default?

**Comment by billziss-gh** [comment]
An investigation revealed that there were two reasons for this problem:

- On "disk" file systems SQL Server issues an `IOCTL_STORAGE_QUERY_PROPERTY / StorageAccessAlignmentProperty` to get the underlying "disk" sector size. Because WinFsp file systems do not live on a real disk, this IOCTL was not being handled.

- SQL Server uses Mount Manager functionality in order to determine the "disk" under the file system. WinFsp did not support the Mount Manager for a number of reasons, but primarily because Mount Manager mountpoints are persistent and WinFsp file system mountpoints are in general transient (and are usually determined by the user or program that launches the file system). 

(In my opinion SQL Server should be using [`FileFsSectorSizeInformation`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-fscc/3e75d97f-1d0b-4e47-b435-73c513837a57)) to the file system to get physical sector size information rather than assuming that there is always a disk mounted by the Mount Manager.)

The current tip (as of 2019/9/8) of the master branch now includes fixes for both problems:

- WinFsp already included a virtual disk device for "disk" file systems and `IOCTL_STORAGE_QUERY_PROPERTY / StorageAccessAlignmentProperty` is now handled there.

- WinFsp now supports the Mount Manager and it does so in a backwards compatible way:
    - If the specified mountpoint is in the syntax `\\.\X:` then the MountManager is used.
    - If the specified mountpoint is in the syntax `X:` then `DefineDosDeviceW` is used (i.e. same as today).
    - If the specified mountpoint is in the syntax `X:\DIR` then a reparse point is used and the file system is mounted as a directory (i.e. same as today).

Some caveats about the Mount Manager:

- It requires Administrator access. This is because `SetVolumeMountPointW` and opening the `\\\\.\\MountPointManager` device require Administrator access. I could have circumvented this in the FSD (since it runs in kernel mode), but WinFsp is not in the business of breaking Windows security.
- It currently only works with drives (`\\.\X:`) and not directories (`\\.\X:\DIR`). If there is sufficient demand I may build support for Mount Manager directories as well.
- Mount Manager mountpoints created by WinFsp are transient. WinFsp takes various steps to ensure that this is the case.
- I expect (but have not tested) that drives created by the Mount Manager are global and are visible across Terminal Services sessions (they go into the `\GLOBAL??` portion of the NT namespace).

----

@lowleveldesign please check if the latest from master fixes the SQL Server problem for you.

**Comment by lowleveldesign** [comment]
@billziss-gh, I tested the version from master and can confirm SQL Server saves the database file with no issues. I tested both the subfolder and the root folder scenario using memfs-x64.exe (`memfs-x64 -m \\.\X:`):

- `SQLCMD.EXE -S "(localdb)\test-fs" -Q "create database test on primary (name=testdb, filename='x:\test\testdb.mdf') log on (name=testdb_lob, filename='C:\logs\testdb.ldf')"`
- `SQLCMD.EXE -S "(localdb)\test-fs" -Q "create database test on primary (name=testdb, filename='x:\testdb.mdf') log on (name=testdb_lob, filename='C:\logs\testdb.ldf')"`

So all looks good! Great thanks for the fix!

PS. As you assumed, the drive is globally visible in all Windows sessions.
